src/node_file.cc

```diff
@@ -3179,44 +3179,38 @@ static void CpSyncCheckPaths(const FunctionCallbackInfo<Value>& args) {
   if (!error_code) {
     // Check if src and dest are identical.
     if (std::filesystem::equivalent(src_path, dest_path)) {
-      std::u8string message =
-          u8"src and dest cannot be the same " + dest_path.u8string();
-      return THROW_ERR_FS_CP_EINVAL(
-          env, reinterpret_cast<const char*>(message.c_str()));
+      std::string message = "src and dest cannot be the same %s";
+      return THROW_ERR_FS_CP_EINVAL(env, message.c_str(), dest_path.string());
     }
 
     const bool dest_is_dir =
         dest_status.type() == std::filesystem::file_type::directory;
 
     if (src_is_dir && !dest_is_dir) {
-      std::u8string message = u8"Cannot overwrite non-directory " +
-                              src_path.u8string() + u8" with directory " +
-                              dest_path.u8string();
+      std::string message =
+          "Cannot overwrite non-directory %s with directory %s";
       return THROW_ERR_FS_CP_DIR_TO_NON_DIR(
-          env, reinterpret_cast<const char*>(message.c_str()));
+          env, message.c_str(), src_path.string(), dest_path.string());
     }
 
     if (!src_is_dir && dest_is_dir) {
-      std::u8string message = u8"Cannot overwrite directory " +
-                              dest_path.u8string() + u8" with non-directory " +
-                              src_path.u8string();
+      std::string message =
+          "Cannot overwrite directory %s with non-directory %s";
       return THROW_ERR_FS_CP_NON_DIR_TO_DIR(
-          env, reinterpret_cast<const char*>(message.c_str()));
+          env, message.c_str(), dest_path.string(), src_path.string());
     }
   }
 
-  std::u8string dest_path_str = dest_path.u8string();
-  std::u8string src_path_str = src_path.u8string();
+  std::string dest_path_str = dest_path.string();
```

Comment:
This might segfault with utf8 chars in windows, since the `src_path` and `dest_path` are constructed from `u8string`.

See my previous findings https://github.com/nodejs/node/pull/54653#issuecomment-2362892778, which identified this was the culprit for the never-passed CI on windows.

Reply:
You're correct, as https://ci.nodejs.org/job/node-test-commit-windows-fanned/66316/ shows this PR does segfault on Windows -.-

Reply:
All `src_path.string()` and `dest_path.string()` need to be avoided to avoid the segfault. I am curious which was the culprit of the hard crash for this issue. Is it something like `dest_path.u8string()`?

Reply:
Ah, I don't think there is an easy way to format u8string : (

Reply:
Would it be easier to format `wstring`? If you use `.native()`, you can branch between Windows and others: https://en.cppreference.com/w/cpp/filesystem/path/native

Reply:
Yeah, I think it can be better. Probably need to do a manual conversion `MultiByteToWideChar` on Windows first

Reply:
Attempt with `.native()`: https://ci.nodejs.org/job/node-test-commit-windows-fanned/66376/

Reply:
> Probably need to do a manual conversion `MultiByteToWideChar` on Windows first

You may be right, but I don't really understand what it would look like (sorry, C++ and Windows dev are not my string suit)

Reply:
I tried something like this, and it looks good from my side
```
#ifdef _WIN32
std::wstring ConvertToWideString(const std::string& str) {
    int size_needed = MultiByteToWideChar(CP_UTF8, 0, &str[0], (int)str.size(), NULL, 0);
    std::wstring wstrTo(size_needed, 0);
    MultiByteToWideChar(CP_UTF8, 0, &str[0], (int)str.size(), &wstrTo[0], size_needed);
    return wstrTo;
}

std::string ConvertWideToUTF8(const std::wstring& wstr)
{
    if (wstr.empty())
        return std::string();

    int size_needed = WideCharToMultiByte(CP_UTF8, 0, &wstr[0], (int)wstr.size(), NULL, 0, NULL, NULL);
    std::string strTo(size_needed, 0);
    WideCharToMultiByte(CP_UTF8, 0, &wstr[0], (int)wstr.size(), &strTo[0], size_needed, NULL, NULL);
    return strTo;
}
#endif
```

And in the implementation, I think it's better to completely branch between windows and others inside `CpSyncCheckPaths`, just like how it's done in `ToNamespacedPath`. 

```
static void CpSyncCheckPaths(const FunctionCallbackInfo<Value>& args) {
  // ...
  
  // to construct a std::filesystem::path
#ifdef _WIN32
  auto src_path = std::filesystem::path(ConvertToWideString(src.ToString()));
  auto  dest_path = std::filesystem::path(ConvertToWideString(dest.ToString()));
#else
  auto src_path = std::filesystem::path(dest.ToStringView());
  auto dest_path = std::filesystem::path(dest.ToStringView());
#endif
  
  // ...
  
  // when throw, for example
#ifdef _WIN32
      return THROW_ERR_FS_CP_EINVAL(
          env, message.c_str(), ConvertWideToUTF8(dest_path.wstring()));
#else
      return THROW_ERR_FS_CP_EINVAL(
          env, message.c_str(), dest_path.string());
#endif
  
   // ...
}
```

Reply:
@jazelly thanks a lot for investigating. Any chance you could share a diff or push a commit?

Reply:
Hope this format would be clearer
https://github.com/nodejs/node/compare/main...jazelly:wstring-windows?expand=1

Reply:
Yay it looks like it's building and passing tests! Thanks!